webapp/src/ts/services/search-telemetry.service.ts

```diff
@@ -0,0 +1,71 @@
+import { Injectable } from '@angular/core';
+
+import { TelemetryService } from '@mm-services/telemetry.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class SearchTelemetryService {
+  private readonly CONTACT_SKIPPED_PROPERTIES = ['_id', '_rev', 'type', 'refid', 'geolocation'];
+  private readonly REPORT_SKIPPED_PROPERTIES = ['_id', '_rev', 'type', 'refid', 'content'];
+
+  constructor(
+    private readonly telemetryService: TelemetryService,
+  ) {}
+
+  public async recordContactSearch(contactDoc: Record<string, any>, search: string) {
+    const matchingProperties = await this.findMatchingProperties(contactDoc, search, this.CONTACT_SKIPPED_PROPERTIES);
+
+    for (const key of matchingProperties) {
+      await this.telemetryService.record(`search_match:contacts_by_freetext:${key}`);
+    }
+  }
+
+  public async recordContactByTypeSearch(contactDoc: Record<string, any>, search: string) {
+    const matchingProperties = await this.findMatchingProperties(contactDoc, search, this.CONTACT_SKIPPED_PROPERTIES);
+
+    for (const key of matchingProperties) {
+      await this.telemetryService.record(`search_match:contacts_by_type_freetext:${key}`);
+    }
+  }
+
+  public async recordReportSearch(reportDoc: Record<string, any>, search: string) {
+    const [matches, fieldsMatches] = await Promise.all([
+      this.findMatchingProperties(reportDoc, search, this.REPORT_SKIPPED_PROPERTIES),
+      this.findMatchingProperties(reportDoc.fields, search, this.REPORT_SKIPPED_PROPERTIES, 'fields'),
+    ]);
+    const matchingProperties = new Set(...matches, ...fieldsMatches);
+
+    for (const key of matchingProperties) {
+      await this.telemetryService.record(`search_match:reports_by_freetext:${key}`);
+    }
+  }
+
+  private async findMatchingProperties(
+    doc: Record<string, any>,
+    search: string,
+    skip: string[],
+    basePropertyPath = '',
+  ) {
+    const matchingProperties = new Set<string>();
+    const colonSearch = search.split(':');
+    if (colonSearch.length > 1) {
+      matchingProperties.add(`${colonSearch[0]}:$value`);
```

Comment:
This is just capturing the fact that someone did a colon search right?  

In this case, should we be adjusting the `_search` value to be `colonSearch[1]`?  OTOH, maybe we should just return with [`${colonSearch[0]}:$value`] and not bother checking the rest of the doc fields...

Reply:
You're right we can just return early and not bother looping over the doc fields